### PR TITLE
feat(sdk): ability to specify a function to fetch the refresh token

### LINKED ISSUE #45286: Ability to pass authentication headers when fetching request tokens in embedding SDK
Some SDK customers has SSO endpoints that authenticates with bearer tokens in the request headers, not cookies. We currently set `credentials: "include"` in the fetch request to get the refresh token, which only works for cookie-based session authentication but not for header-based bearer tokens. We should expose an API to allow SDK customers to customize authentication headers. 

We currently use the `refreshTokenAsync` thunk to get the refresh token from the customer's SSO endpoint: https://github.com/metabase/metabase/blob/master/enterprise/frontend/src/embedding-sdk/store/reducer.ts#L49-L58

We could implement this in two ways:

Option 1: Specify a function to get the refresh token.

```ts
const config = {
  async getAuthRefreshToken() {
    // customer calls fetch() or calls a function they have to get a refresh token
  }
}

<MetabaseProvider config={config} />
```

Option 2: Specify a function to get the auth header, which we internally pass to `fetch()` and remove `credentials: "include"`

```ts
const config = {
  async getAuthHeaders() {
    return { "Authorization": "<whatever>" }
  }
}

<MetabaseProvider config={config} />
```

### PR DESCRIPTION
**Pull request by heypoom**
Closes https://github.com/metabase/metabase/issues/45286

### Description

Allows SDK users to pass their own function to refresh the authentication token. This allows them to specify their own implementation of querying the refresh token, such as for token-based authentication (as opposed to session-based)

```ts
// This is our default implementation.
async function fetchRefreshToken(url: string) {
  const response = await fetch(url, {
    method: "GET",
    credentials: "include",
  });

  return await response.json();
}

const config = { fetchRefreshToken }
```

### How to verify

TBA

### Demo

TBA

### Checklist

- [ ] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: enterprise/frontend/src/embedding-sdk/README.md [hunk 1]
```diff
@@ -688,6 +688,27 @@ useEffect(() => {
...
+// You can customize this function to fit your needs, such as adding headers or excluding cookies.
+// The function must return a JWT token object in the format of {id: string, exp: number}
+async function fetchRefreshToken(url) {
+  const response = await fetch(url, {
+    method: "GET",
+    credentials: "include",
+  });
+
+  return await response.json();
+}
```
Comment (sbrudz) [inline enterprise/frontend/src/embedding-sdk/README.md:706]: How should we handle errors?  throw?

For the case where the user hasn't logged in yet (and doesn't have a bearer token), can we return an empty or null payload?

Comment (heypoom) [reply enterprise/frontend/src/embedding-sdk/README.md:706]: @sbrudz Thank you for the fast feedback! Yes, you can return `null` when there is no bearer tokens, and you can throw the error in this function. I'll update the docs and inline comments to clarify this 🙏🏻

File: enterprise/frontend/src/embedding-sdk/store/reducer.ts [hunk 2]
```diff
@@ -48,12 +55,27 @@ export const getOrRefreshSession = createAsyncThunk(
...
+    } catch (errorCause) {
+      // As this function can be supplied by the SDK user,
+      // we have to handle possible errors in refreshing the token.
+      const error = new Error(t`failed to refresh the auth token`);
+      error.cause = errorCause;
+
+      setLoginStatus({ status: "error", error });
+
+      return null;
+    }
```
Comment (heypoom) [inline enterprise/frontend/src/embedding-sdk/store/reducer.ts:78]: @oisincoveney I'm not sure if using `setLoginStatus` makes sense here - in this case this happens when we fail to refresh the authentication token, so I think it makes sense to set the login status to "error" here. As this function can be specified by the SDK user, there's always a possibility of error.

File: enterprise/frontend/src/embedding-sdk/README.md (hunk 1 above)
Comment (sbrudz) [reply enterprise/frontend/src/embedding-sdk/README.md]: Great! Thank you!

**Review by deniskaber — APPROVED** [review]
Looks good!

**Comment by heypoom** [comment]
PS. Looks like markdown-lint-check is failing due to [dead links in our docs](https://github.com/metabase/metabase/actions/runs/9870440055/job/27256186793?pr=45313), e.g. https://logging.apache.org/log4j/2.x/ and https://directory.apache.org/studio/. This seems to be caused by a downtime in Apache's directory servers, so the linter thinks they are dead links.

**Pull request merged**